src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,93 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Net.Sockets
+{
+    internal class IOControlKeepAlive
+    {
+        private const uint OnOff = 1u;
+        private const uint DefaultTimeMs = 7200000u;
+        private const uint DefaultIntervalMs = 1000u;
+        private static readonly ConditionalWeakTable<Socket, IOControlKeepAlive> SocketKeepAliveCwt;
+        private static readonly int UIntSize;
+
+        private uint timeMs;
+        private uint intervalMs;
+
+        static IOControlKeepAlive()
+        {
+            SocketKeepAliveCwt = new ConditionalWeakTable<Socket, IOControlKeepAlive>();
+            UIntSize = Marshal.SizeOf<uint>();
+        }
+
+        public static bool TryGet(Socket socket, SocketOptionLevel optionLevel, SocketOptionName optionName, ref int optionValue)
+        {
+            if (!IsNeeded(optionLevel, optionName))
+                return false;
+                
+            IOControlKeepAlive ioControlKeepAlive;
+            SocketKeepAliveCwt.TryGetValue(socket, out ioControlKeepAlive);
+            optionValue = optionName == SocketOptionName.TcpKeepAliveTime ?
+                ToS(ioControlKeepAlive?.timeMs ?? DefaultTimeMs) :
+                ToS(ioControlKeepAlive?.intervalMs ?? DefaultIntervalMs);
+            return true;
+        }
+
+        public static bool TrySet(Socket socket, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
+        {
+            if (!IsNeeded(optionLevel, optionName))
+                return false;
+
+            var ioControlKeepAlive = SocketKeepAliveCwt.GetOrCreateValue(socket);
+            if (optionName == SocketOptionName.TcpKeepAliveTime)
+                ioControlKeepAlive.timeMs = ToMs(optionValue);
+            else
+                ioControlKeepAlive.intervalMs = ToMs(optionValue);
+            socket.IOControl(IOControlCode.KeepAliveValues, ioControlKeepAlive.ToByteArray(), null);
+            return true;
+        }
+
+        public IOControlKeepAlive()
+        {
+            timeMs = DefaultTimeMs;
+            intervalMs = DefaultIntervalMs;
+        }
+
+        private static bool IsNeeded(SocketOptionLevel optionLevel, SocketOptionName optionName)
+        {
+            return IsWindowsLessThan10Version1709() &&
+                optionLevel == SocketOptionLevel.Tcp &&
+                (optionName == SocketOptionName.TcpKeepAliveTime || optionName == SocketOptionName.TcpKeepAliveInterval);
+        }
+
+        private static bool IsWindowsLessThan10Version1709()
+        {
+            return RuntimeInformation.IsOSPlatform(OSPlatform.Windows) &&
+                Interop.NtDll.RtlGetVersionEx(out Interop.NtDll.RTL_OSVERSIONINFOEX os) == 0 &&
+                (os.dwMajorVersion < 10 || (os.dwMajorVersion == 10 && os.dwMinorVersion == 0 && os.dwBuildNumber < 16299));
+        }
```

Comment:
This should be cached rather than accessing it every time.

Also, we generally like to avoid doing explicit version checks as a proxy for whether certain features are supported, and instead prefer to simply see whether that feature is available.  Could we instead just do something like:
```C#
private static readonly bool s_supportsSetSocketOption = SupportsSetSocketOption();
private static bool SupportsSocketOption()
{
    using (var socket = new Socket(...))
    {
        return SocketPal.SetSockOpt(socket, ...) == SocketError.Success;
    }
}
```
and effectively let the OS tell us whether the option is supported or not?

Reply:
Is there an implicit cast from Socket to SafeSocketHandle?